src/OSPSuite.R/Services/ConcurrentSimulationRunner.cs

```diff
@@ -33,82 +74,140 @@ public interface IConcurrentSimulationRunner : IDisposable
       /// </summary>
-      /// <param name="simulationBatch"></param>
-      void AddSimulationBatch(SimulationBatch simulationBatch);
+      /// <param name="simulationWithBatchOptions">the options to run the batch</param>
+      void AddSimulationBatchOption(SettingsForConcurrentRunSimulationBatch simulationWithBatchOptions);
 
+      /// <summary>
+      /// Clear all data for freshly start
+      /// </summary>
       void Clear();
 
       /// <summary>
       ///    Runs all preset settings concurrently
       /// </summary>
       /// <returns></returns>
-      SimulationResults[] RunConcurrently();
+      ConcurrentSimulationResults[] RunConcurrently();
+
+      /// <summary>
+      /// After initialization phase, run all simulations or simulationBatches Async
+      /// </summary>
+      /// <returns></returns>
+      Task<IEnumerable<ConcurrentSimulationResults>> RunConcurrentlyAsync();
    }
 
    public class ConcurrentSimulationRunner : IConcurrentSimulationRunner
    {
-      private readonly IConcurrencyManager _concurrentManager;
+      private readonly IConcurrencyManager _concurrencyManager;
 
       public ConcurrentSimulationRunner(IConcurrencyManager concurrentManager)
       {
-         _concurrentManager = concurrentManager;
+         _concurrencyManager = concurrentManager;
       }
 
       public SimulationRunOptions SimulationRunOptions { get; set; }
 
       public int NumberOfCores { get; set; }
 
       private readonly List<IModelCoreSimulation> _simulations = new List<IModelCoreSimulation>();
-      private readonly List<SimulationBatch> _simulationBatches = new List<SimulationBatch>();
+      private readonly List<SettingsForConcurrentRunSimulationBatch> _listOfSettingsForConcurrentRunSimulationBatch = new List<SettingsForConcurrentRunSimulationBatch>();
       private CancellationTokenSource _cancellationTokenSource;
 
       public void AddSimulation(IModelCoreSimulation simulation)
       {
          _simulations.Add(simulation);
       }
 
-      public void AddSimulationBatch(SimulationBatch simulationBatch)
+      public void AddSimulationBatchOption(SettingsForConcurrentRunSimulationBatch settings)
       {
-         _simulationBatches.Add(simulationBatch);
+         _listOfSettingsForConcurrentRunSimulationBatch.Add(settings);
       }
 
       public void Clear()
       {
          _simulations.Clear();
-         _simulationBatches.Clear();
-         if (_cancellationTokenSource != null)
-            _cancellationTokenSource.Cancel();
+         _listOfSettingsForConcurrentRunSimulationBatch.Clear();
+         _cancellationTokenSource?.Cancel();
       }
 
-      public SimulationResults[] RunConcurrently()
+      private Task initializeBatches()
+      {
+          return _concurrencyManager.RunAsync
+         (
+            NumberOfCores,
+            _cancellationTokenSource.Token,
+            _listOfSettingsForConcurrentRunSimulationBatch.SelectMany
```

Comment:
@abdelr A quick question. Why is it not  a `Select`?

Reply:
@msevestre I have just added a comment to explain on the code, so it will be there in three months when we do not remember why anymore. I think we could maybe do it on a single select with the sum of all needed batches over the different simulations but still we will need to know for each iteration what settings object it refers to so I would say it is maybe easier and clearer with SelectMany on the list of simulations, Select on the list of batches for the given simulation.